Separating the functions that were used in multiple files

iifawzi opened the pull request:
### Description

This Pull Request introduces some of the utilities that was being used in the customizers classes (ImageCustom, fontCustom, svgCustom). I will add more details about the changes in separate reviews. 
Utils main folder, will have the `CustomFont`, `CustomImage` and `CustomizeSvg` classes.. for that reason I put those utils in `tools` folder.


If this PR should be separated into multiple PRs, let me know.. I just didn't want to spam the PRs by making a separate PR for each tool

iifawzi commented on src/utils/tools/addConfigFile.ts:
```diff
@@ -0,0 +1,11 @@
+    throw new Error(`An error occured while adding the config file: ${err}`)
+  }
+}
+
```
- `AddConfigFile` method was being used in `customImage` class.
```JS 
fs.writeFileSync(
      `${tempDir}dist_${this.timestamp}/customizationConfig.json`,
      JSON.stringify(this.payload),
      (err) => {
        if (err) {
          console.log(err);
        }
      }
    );
```
I've changed the synchronization call to an Async call and used await on it, by this way, it will prevent the entire event loop from being blocked, only the context that encounters await will get blocked ([async+await == sync?](https://softwareengineering.stackexchange.com/questions/183576/asyncawait-sync))

iifawzi commented on src/utils/tools/isProduction.ts:
```diff
@@ -0,0 +1,11 @@
+    return true
+  }
+  return false
+}
```
This was being used in the `postAnalyticsData`, no changes was made to the code, I've only put it in a separate files for a separation of concerns, since the process of posting analytics data will be placed in a the service folder that will be only responsible of dealing with the analytics collection in mongo (This will not affect the functionality of `v1` API).

iifawzi commented on src/utils/tools/serializer.ts:
```diff
@@ -0,0 +1,18 @@
+import { ExportType } from 'contracts/types'
+const serializer = (payload: any, format: ExportType, customized: boolean = false) => {
+  const data = {
+    customizationConfig: {},
+    customized: customized,
+    timestamp: new Date(),
+    icons: payload.icons,
+    format: format,
+    location: undefined
+  }
+
+  if (customized) {
+    data.customizationConfig = payload.customizationConfig
+  }
+  return data
+}
```
This serializer was being used in `postAnalyticsData`: 
```JS
const serializer = (payload, format, customized) => {

  const data = {
    timestamp: new Date(),
    location: undefined,
    format: format,
    customized: customized,
  }

  if (customized) {
    data.customizationConfig = payload.customizationConfig
    data.customizationConfig.size = payload.exportSize
  }
  data.icons = payload.icons

  return data

}

```
For the same reason I mentioned above, I put it in a separate file.

I've removed this line, since it doesn't add any additional info I think, the exportSize will be stored in the CustomizationConfig object, why would we need to store it again with `size` property? I might be missing something, let me know if it has different use or meaning. 
```JS 
 data.customizationConfig.size = payload.exportSize
```

Also, it seems like `location` is always set to undefined, can we remove it?

kaybhutani replied:
yes, lets remove location. We thought of adding it but with gdpr and stuff looks like we wont be able to

iifawzi replied:
I figured out that I was mistaken, the `size` property isn't already in the `customizationConfig`. I will add it again.

iifawzi commented on src/utils/tools/svgToPng.ts:
```diff
@@ -0,0 +1,18 @@
+import pify from 'pify'
+import svg2img from 'svg2img'
+const psvg2img = pify(svg2img)
+
+async function svgToPng (pngSize: string, svgPath: string) {
+  try {
+    const svgOptions = {
+      width: pngSize,
+      height: pngSize
+    }
+    const buffer = await psvg2img(svgPath, svgOptions)
+    return buffer
+  } catch (err) {
+    throw new Error(`An error occured while converting the svg to png: ${err}`)
+  }
+}
+
```
This is the same `svgToPng` utility: 

```JS
const fs = require('fs');
const svg2img = require('svg2img');
const tempDir = `${__dirname}/../temp/`;

convertSvgToPng = (iconName, pngSize, theme = 'svg') => {
  return new Promise((resolve, reject) => {
    if (!fs.existsSync(tempDir)) {
      fs.mkdirSync(tempDir);
    }
    svg2img(`${__dirname}/../${theme}/${iconName}.svg`, { width: pngSize, height: pngSize }, (error, buffer) => {
      if (error) {
        return reject();
      }
      fs.writeFileSync(`${__dirname}/../temp/${iconName}_${pngSize}.png`, buffer);
      resolve();
    });
  });
};

module.exports = convertSvgToPng;

```

The changes: 
1) Using a promise style version (converted by `pify`). 
2) Passing the `svgPath` as parameter to adapt any changes easily. 
3) Making it only responsible about converting the `SVG` and return the buffer, the write process  will be called from the controllers.

iifawzi commented on src/utils/tools/zipFolder.ts:
```diff
@@ -0,0 +1,14 @@
+// promise-style version of zip-folder
+const pzipFolder = pify(zipFolderPackage)
+
+async function zipFolder (folderPath: string, outputPath: string) {
+  try {
+    await pzipFolder(folderPath, outputPath)
+  } catch (err) {
+    throw new Error(`Some error occurred while generating the zip file: ${err}`)
+  }
+}
+
```
The zipping process was used in `CustomImage` and `customFont` classes, only moved to a separate tool and used a Promise style version.

kaybhutani replied:
not that it makes any diff but lets make everything consistent, i saw arrow functions before, maybe use that only from now

iifawzi replied:
You're right, I will stick with the arrow functions, gonna make the changes.

iifawzi replied:
Resolved.

iifawzi commented:
I'm thinking of moving all the errors to an `errors` object as following: 
```JS
export errors = {
FILE_NOT_FOUND: "File is not found, please .... ",
ERROR_ZIPPING_FILE: "An error occurred while zipping the folder, please ...." 
}
```
and then those errors can be used as following: 
```JS
errors.FILE_NOT_FOUND
```
what do you think? this will make the code more cleaner and will help us having all the errors in one place which will make the modification easier.

kaybhutani commented on src/utils/tools/serializer.ts:
```diff
@@ -0,0 +1,17 @@
+import { ExportType } from 'contracts/types'
+const serializer = (payload: any, format: ExportType, customized: boolean = false) => {
```
can we add a type for payload too?
i guess icons will be always there and customizationConfig in case of images only, so maybe something like

```
interface Payload {
icons: string[],
customizationConfig?: interfaceForThat
}```

iifawzi replied:
Sure we can it will look as following, the most of the properties will be optional for now, since the schema isn't consistent. 
```TS
interface Payload {
icons: string[],
exportAs?: ExportType,
exportSize?: number,
customizationConfig?: interfaceForThat,
}```
```
Maybe later in `v2` we can make it consistent.

kaybhutani commented on src/utils/tools/serializer.ts:
```diff
@@ -0,0 +1,17 @@
+import { ExportType } from 'contracts/types'
+const serializer = (payload: any, format: ExportType, customized: boolean = false) => {
+  const data = {
```
lets make a type for this data too, since this is going into our database, i think we should have proper types and models for this now

iifawzi replied:
The database model of the database was suppose to be made when we attempt to create the model (probably using mongoose ) and the service that will insert the data. For that reason I decided to keep it for that PR. If you do prefer adding it right now, I will do.

kaybhutani commented on tsconfig.json:
```diff
@@ -10,6 +10,11 @@
     "esModuleInterop": true,                        /* Enables emit interoperability between CommonJS and ES Modules via creation of namespace objects for all imports. Implies 'allowSyntheticDefaultImports'. */
     /* Advanced Options */
     "skipLibCheck": true,                           /* Skip type checking of declaration files. */
+<<<<<<< HEAD
```
you missed resolving the conflicts i guess @iifawzi

iifawzi replied:
Resolved.

iifawzi commented:
I've added all the needed types, could you re-check? wouldn't be amazed if I missed something.  @kbhutani0001 

I've added the payloads and configurations interfaces in the `types.ts` file, since they will be also used in controllers and the customizers classes. I'm not really sure what's better, putting all public interfaces in a single file? `types.ts` or making a bunch of separated files in the `contracts` folder where each contain a specific and related interfaces?
